# **ECDC_COVID_19:DATASET DESCRIPTION**

* dateRep: The date of the reported data

* day: The day of the report (integer, e.g., 1-31).

* month: The month of the report (integer, e.g., 1-12).

* year: The year of the report (integer, e.g., 2020).

* cases: The number of new confirmed cases reported on this date.

* deaths: The number of new deaths reported on this date.

* countriesAndTerritories: The name of the country or territory to which the data belongs.

* geoId: A short geographical identifier (code) for the country or territory.

* countryterritoryCode



In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv('ECDC_COVID_19.csv')

In [3]:
df

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0     2020-04-08    8      4  2020     30       4             Afghanistan   
1     2020-04-07    7      4  2020     38       0             Afghanistan   
2     2020-04-06    6      4  2020     29       2             Afghanistan   
3     2020-04-05    5      4  2020     35       1             Afghanistan   
4     2020-04-04    4      4  2020      0       0             Afghanistan   
...          ...  ...    ...   ...    ...     ...                     ...   
9508  2020-03-25   25      3  2020      0       0                Zimbabwe   
9509  2020-03-24   24      3  2020      0       1                Zimbabwe   
9510  2020-03-23   23      3  2020      0       0                Zimbabwe   
9511  2020-03-22   22      3  2020      1       0                Zimbabwe   
9512  2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018  
0       AF                  AFG   37172386.0  
1       AF                  AFG   37172386.0  
2       AF                  AFG   37172386.0  
3       AF                  AFG   37172386.0  
4       AF                  AFG   37172386.0  
...    ...                  ...          ...  
9508    ZW                  ZWE   14439018.0  
9509    ZW                  ZWE   14439018.0  
9510    ZW                  ZWE   14439018.0  
9511    ZW                  ZWE   14439018.0  
9512    ZW                  ZWE   14439018.0  

[9513 rows x 10 columns]

In [4]:
df.ndim

2

In [5]:
df.head(10)

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0  2020-04-08    8      4  2020     30       4             Afghanistan    AF   
1  2020-04-07    7      4  2020     38       0             Afghanistan    AF   
2  2020-04-06    6      4  2020     29       2             Afghanistan    AF   
3  2020-04-05    5      4  2020     35       1             Afghanistan    AF   
4  2020-04-04    4      4  2020      0       0             Afghanistan    AF   
5  2020-04-03    3      4  2020     43       0             Afghanistan    AF   
6  2020-04-02    2      4  2020     26       0             Afghanistan    AF   
7  2020-04-01    1      4  2020     25       0             Afghanistan    AF   
8  2020-03-31   31      3  2020     27       0             Afghanistan    AF   
9  2020-03-30   30      3  2020      8       1             Afghanistan    AF   

  countryterritoryCode  popData2018  
0                  AFG   37172386.0  
1                  AFG   37172386.0  
2                  AFG   37172386.0  
3                  AFG   37172386.0  
4                  AFG   37172386.0  
5                  AFG   37172386.0  
6                  AFG   37172386.0  
7                  AFG   37172386.0  
8                  AFG   37172386.0  
9                  AFG   37172386.0

In [6]:
df.tail(10)

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
9503  2020-03-30   30      3  2020      0       0                Zimbabwe   
9504  2020-03-29   29      3  2020      2       0                Zimbabwe   
9505  2020-03-28   28      3  2020      2       0                Zimbabwe   
9506  2020-03-27   27      3  2020      0       0                Zimbabwe   
9507  2020-03-26   26      3  2020      1       0                Zimbabwe   
9508  2020-03-25   25      3  2020      0       0                Zimbabwe   
9509  2020-03-24   24      3  2020      0       1                Zimbabwe   
9510  2020-03-23   23      3  2020      0       0                Zimbabwe   
9511  2020-03-22   22      3  2020      1       0                Zimbabwe   
9512  2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018  
9503    ZW                  ZWE   14439018.0  
9504    ZW                  ZWE   14439018.0  
9505    ZW                  ZWE   14439018.0  
9506    ZW                  ZWE   14439018.0  
9507    ZW                  ZWE   14439018.0  
9508    ZW                  ZWE   14439018.0  
9509    ZW                  ZWE   14439018.0  
9510    ZW                  ZWE   14439018.0  
9511    ZW                  ZWE   14439018.0  
9512    ZW                  ZWE   14439018.0

In [7]:
df.size

95130

In [8]:
df.index

RangeIndex(start=0, stop=9513, step=1)

In [9]:
df.shape

(9513, 10)

## Information of the dataset

* Columns= 10 
* Rows= 9513
* Data types:
    - float = 1
    - integer = 5
    - object = 4
* Occur Null values
    - geoId 
    - countryterritoryCode
    - popData2018

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9513 entries, 0 to 9512
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   dateRep                  9513 non-null   object 
 1   day                      9513 non-null   int64  
 2   month                    9513 non-null   int64  
 3   year                     9513 non-null   int64  
 4   cases                    9513 non-null   int64  
 5   deaths                   9513 non-null   int64  
 6   countriesAndTerritories  9513 non-null   object 
 7   geoId                    9488 non-null   object 
 8   countryterritoryCode     9324 non-null   object 
 9   popData2018              9370 non-null   float64
dtypes: float64(1), int64(5), object(4)
memory usage: 743.3+ KB


In [11]:
df.describe()

day        month         year         cases       deaths  \
count  9513.000000  9513.000000  9513.000000   9513.000000  9513.000000   
mean     15.548828     2.591822  2019.992957    146.314517     8.564911   
std       9.427045     1.284901     0.083631   1141.114073    71.611262   
min       1.000000     1.000000  2019.000000     -9.000000     0.000000   
25%       7.000000     2.000000  2020.000000      0.000000     0.000000   
50%      16.000000     3.000000  2020.000000      0.000000     0.000000   
75%      24.000000     3.000000  2020.000000     13.000000     0.000000   
max      31.000000    12.000000  2020.000000  34272.000000  2004.000000   

        popData2018  
count  9.370000e+03  
mean   6.431196e+07  
std    2.015642e+08  
min    1.000000e+03  
25%    3.545883e+06  
50%    1.062716e+07  
75%    4.272314e+07  
max    1.392730e+09

In [12]:
df.describe(include='all')

dateRep          day        month         year         cases  \
count         9513  9513.000000  9513.000000  9513.000000   9513.000000   
unique         100          NaN          NaN          NaN           NaN   
top     2020-04-08          NaN          NaN          NaN           NaN   
freq           203          NaN          NaN          NaN           NaN   
mean           NaN    15.548828     2.591822  2019.992957    146.314517   
std            NaN     9.427045     1.284901     0.083631   1141.114073   
min            NaN     1.000000     1.000000  2019.000000     -9.000000   
25%            NaN     7.000000     2.000000  2020.000000      0.000000   
50%            NaN    16.000000     3.000000  2020.000000      0.000000   
75%            NaN    24.000000     3.000000  2020.000000     13.000000   
max            NaN    31.000000    12.000000  2020.000000  34272.000000   

             deaths countriesAndTerritories geoId countryterritoryCode  \
count   9513.000000                    9513  9488                 9324   
unique          NaN                     204   203                  199   
top             NaN                   Italy    KR                  SGP   
freq            NaN                     100   100                  100   
mean       8.564911                     NaN   NaN                  NaN   
std       71.611262                     NaN   NaN                  NaN   
min        0.000000                     NaN   NaN                  NaN   
25%        0.000000                     NaN   NaN                  NaN   
50%        0.000000                     NaN   NaN                  NaN   
75%        0.000000                     NaN   NaN                  NaN   
max     2004.000000                     NaN   NaN                  NaN   

         popData2018  
count   9.370000e+03  
unique           NaN  
top              NaN  
freq             NaN  
mean    6.431196e+07  
std     2.015642e+08  
min     1.000000e+03  
25%     3.545883e+06  
50%     1.062716e+07  
75%     4.272314e+07  
max     1.392730e+09

In [13]:
df.describe(include='object')

dateRep countriesAndTerritories geoId countryterritoryCode
count         9513                    9513  9488                 9324
unique         100                     204   203                  199
top     2020-04-08                   Italy    KR                  SGP
freq           203                     100   100                  100

In [14]:
df.isnull().sum()

dateRep                      0
day                          0
month                        0
year                         0
cases                        0
deaths                       0
countriesAndTerritories      0
geoId                       25
countryterritoryCode       189
popData2018                143
dtype: int64

In [15]:
df.nunique()

dateRep                    100
day                         31
month                        5
year                         2
cases                      728
deaths                     222
countriesAndTerritories    204
geoId                      203
countryterritoryCode       199
popData2018                199
dtype: int64

In [16]:
df.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
9508    False
9509    False
9510    False
9511    False
9512    False
Length: 9513, dtype: bool

### No duplicate Values

In [17]:
df.duplicated().sum()

0

In [18]:
df.columns

Index(['dateRep', 'day', 'month', 'year', 'cases', 'deaths',
       'countriesAndTerritories', 'geoId', 'countryterritoryCode',
       'popData2018'],
      dtype='object')

In [19]:
df['day'].unique()

array([ 8,  7,  6,  5,  4,  3,  2,  1, 31, 30, 29, 28, 27, 26, 25, 24, 23,
       22, 21, 20, 19, 18, 17, 16, 15, 11, 14, 13, 12, 10,  9],
      dtype=int64)

In [20]:
df.value_counts()

dateRep     day  month  year  cases  deaths  countriesAndTerritories   geoId  countryterritoryCode  popData2018
2019-12-31  31   12     2019  0      0       Afghanistan               AF     AFG                   37172386.0     1
2020-03-23  23   3      2020  38     0       Lithuania                 LT     LTU                   2789533.0      1
                              122    0       Thailand                  TH     THA                   69428524.0     1
                              123    6       Malaysia                  MY     MYS                   31528585.0     1
                              128    0       Luxembourg                LU     LUX                   607728.0       1
                                                                                                                  ..
2020-02-16  16   2      2020  0      0       New_Zealand               NZ     NZL                   4885500.0      1
                                             Nigeria                 

In [21]:
df.dtypes

dateRep                     object
day                          int64
month                        int64
year                         int64
cases                        int64
deaths                       int64
countriesAndTerritories     object
geoId                       object
countryterritoryCode        object
popData2018                float64
dtype: object

In [22]:
df.select_dtypes('object')

dateRep countriesAndTerritories geoId countryterritoryCode
0     2020-04-08             Afghanistan    AF                  AFG
1     2020-04-07             Afghanistan    AF                  AFG
2     2020-04-06             Afghanistan    AF                  AFG
3     2020-04-05             Afghanistan    AF                  AFG
4     2020-04-04             Afghanistan    AF                  AFG
...          ...                     ...   ...                  ...
9508  2020-03-25                Zimbabwe    ZW                  ZWE
9509  2020-03-24                Zimbabwe    ZW                  ZWE
9510  2020-03-23                Zimbabwe    ZW                  ZWE
9511  2020-03-22                Zimbabwe    ZW                  ZWE
9512  2020-03-21                Zimbabwe    ZW                  ZWE

[9513 rows x 4 columns]

# Feature Engineering

## Covert datatype
dateRep (object) - datetime

In [23]:
df['dateRep']=df['dateRep'].astype('datetime64[ns]')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9513 entries, 0 to 9512
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   dateRep                  9513 non-null   datetime64[ns]
 1   day                      9513 non-null   int64         
 2   month                    9513 non-null   int64         
 3   year                     9513 non-null   int64         
 4   cases                    9513 non-null   int64         
 5   deaths                   9513 non-null   int64         
 6   countriesAndTerritories  9513 non-null   object        
 7   geoId                    9488 non-null   object        
 8   countryterritoryCode     9324 non-null   object        
 9   popData2018              9370 non-null   float64       
dtypes: datetime64[ns](1), float64(1), int64(5), object(3)
memory usage: 743.3+ KB


In [24]:
df['month_name']=df['dateRep'].dt.month_name()

In [25]:
df['day_name']=df['dateRep'].dt.day_name()

In [26]:
df.head()

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0 2020-04-08    8      4  2020     30       4             Afghanistan    AF   
1 2020-04-07    7      4  2020     38       0             Afghanistan    AF   
2 2020-04-06    6      4  2020     29       2             Afghanistan    AF   
3 2020-04-05    5      4  2020     35       1             Afghanistan    AF   
4 2020-04-04    4      4  2020      0       0             Afghanistan    AF   

  countryterritoryCode  popData2018 month_name   day_name  
0                  AFG   37172386.0      April  Wednesday  
1                  AFG   37172386.0      April    Tuesday  
2                  AFG   37172386.0      April     Monday  
3                  AFG   37172386.0      April     Sunday  
4                  AFG   37172386.0      April   Saturday

### **isin()**

it is used to retreive data is in the dataset.
isin() is a method in Pandas used to filter data. It checks whether each element in a DataFrame or Series is contained in a provided list-like object (like a list, set, or Series). This method is often used for subsetting data based on multiple conditions.

In [27]:
df['countriesAndTerritories'].unique()

array(['Afghanistan', 'Albania', 'Andorra', 'Algeria', 'Angola',
       'Anguilla', 'Antigua_and_Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin',
       'Bermuda', 'Bhutan', 'Bolivia',
       'Bonaire, Saint Eustatius and Saba', 'Bosnia_and_Herzegovina',
       'Botswana', 'Brazil', 'British_Virgin_Islands',
       'Brunei_Darussalam', 'Bulgaria', 'Burkina_Faso', 'Burundi',
       'Cambodia', 'Cameroon', 'Canada', 'Cape_Verde',
       'Cases_on_an_international_conveyance_Japan', 'Cayman_Islands',
       'Central_African_Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Congo', 'Costa_Rica', 'Cote_dIvoire', 'Croatia', 'Cuba',
       'Curaçao', 'Cyprus', 'Czechia',
       'Democratic_Republic_of_the_Congo', 'Denmark', 'Djibouti',
       'Dominica', 'Dominican_Republic', 'Ecuador', 'Egypt',
       'El_Salvador', 'Equatorial_Guinea', 'Eritrea', 'Es

###  Q) filter the dataset to include only rows where the countriesandterritories column is either united_states_of_america or india?

In [28]:
usa_india=df[df['countriesAndTerritories'].isin(['United_States_of_America','India'])]
usa_india

dateRep  day  month  year  cases  deaths   countriesAndTerritories  \
4000 2020-04-08    8      4  2020    773      35                     India   
4001 2020-04-07    7      4  2020    354       5                     India   
4002 2020-04-06    6      4  2020    693      32                     India   
4003 2020-04-05    5      4  2020    472       9                     India   
4004 2020-04-04    4      4  2020    601      12                     India   
...         ...  ...    ...   ...    ...     ...                       ...   
9283 2020-01-04    4      1  2020      0       0  United_States_of_America   
9284 2020-01-03    3      1  2020      0       0  United_States_of_America   
9285 2020-01-02    2      1  2020      0       0  United_States_of_America   
9286 2020-01-01    1      1  2020      0       0  United_States_of_America   
9287 2019-12-31   31     12  2019      0       0  United_States_of_America   

     geoId countryterritoryCode   popData2018 month_name   day_name  
4000    IN                  IND  1.352617e+09      April  Wednesday  
4001    IN                  IND  1.352617e+09      April    Tuesday  
4002    IN                  IND  1.352617e+09      April     Monday  
4003    IN                  IND  1.352617e+09      April     Sunday  
4004    IN                  IND  1.352617e+09      April   Saturday  
...    ...                  ...           ...        ...        ...  
9283    US                  USA  3.271674e+08    January   Saturday  
9284    US                  USA  3.271674e+08    January     Friday  
9285    US                  USA  3.271674e+08    January   Thursday  
9286    US                  USA  3.271674e+08    January  Wednesday  
9287    US                  USA  3.271674e+08   December    Tuesday  

[199 rows x 12 columns]

In [29]:
df['month_name'].unique()

array(['April', 'March', 'February', 'January', 'December'], dtype=object)

### Q) filter the dataframe to include only the rows where the month column is either january or february

In [30]:
jan_feb=df[df['month_name'].isin(['January','February'])]
jan_feb

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
29   2020-02-29   29      2  2020      0       0             Afghanistan   
30   2020-02-28   28      2  2020      0       0             Afghanistan   
31   2020-02-27   27      2  2020      0       0             Afghanistan   
32   2020-02-26   26      2  2020      0       0             Afghanistan   
33   2020-02-25   25      2  2020      1       0             Afghanistan   
...         ...  ...    ...   ...    ...     ...                     ...   
9467 2020-01-05    5      1  2020      0       0                 Vietnam   
9468 2020-01-04    4      1  2020      0       0                 Vietnam   
9469 2020-01-03    3      1  2020      0       0                 Vietnam   
9470 2020-01-02    2      1  2020      0       0                 Vietnam   
9471 2020-01-01    1      1  2020      0       0                 Vietnam   

     geoId countryterritoryCode  popData2018 month_name   day_name  
29      AF                  AFG   37172386.0   February   Saturday  
30      AF                  AFG   37172386.0   February     Friday  
31      AF                  AFG   37172386.0   February   Thursday  
32      AF                  AFG   37172386.0   February  Wednesday  
33      AF                  AFG   37172386.0   February    Tuesday  
...    ...                  ...          ...        ...        ...  
9467    VN                  VNM   95540395.0    January     Sunday  
9468    VN                  VNM   95540395.0    January   Saturday  
9469    VN                  VNM   95540395.0    January     Friday  
9470    VN                  VNM   95540395.0    January   Thursday  
9471    VN                  VNM   95540395.0    January  Wednesday  

[4020 rows x 12 columns]

### Q) select the rows where geoid is not in a list of excluded codes (excluded_codes=['US','IN','BR'])?

In [31]:
excluded_codes=['US','IN','BR']

In [32]:
rest_codes=df[~df['geoId'].isin(['excluded_codes'])]
# '~' denotes 'not in'
rest_codes

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0    2020-04-08    8      4  2020     30       4             Afghanistan   
1    2020-04-07    7      4  2020     38       0             Afghanistan   
2    2020-04-06    6      4  2020     29       2             Afghanistan   
3    2020-04-05    5      4  2020     35       1             Afghanistan   
4    2020-04-04    4      4  2020      0       0             Afghanistan   
...         ...  ...    ...   ...    ...     ...                     ...   
9508 2020-03-25   25      3  2020      0       0                Zimbabwe   
9509 2020-03-24   24      3  2020      0       1                Zimbabwe   
9510 2020-03-23   23      3  2020      0       0                Zimbabwe   
9511 2020-03-22   22      3  2020      1       0                Zimbabwe   
9512 2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018 month_name   day_name  
0       AF                  AFG   37172386.0      April  Wednesday  
1       AF                  AFG   37172386.0      April    Tuesday  
2       AF                  AFG   37172386.0      April     Monday  
3       AF                  AFG   37172386.0      April     Sunday  
4       AF                  AFG   37172386.0      April   Saturday  
...    ...                  ...          ...        ...        ...  
9508    ZW                  ZWE   14439018.0      March  Wednesday  
9509    ZW                  ZWE   14439018.0      March    Tuesday  
9510    ZW                  ZWE   14439018.0      March     Monday  
9511    ZW                  ZWE   14439018.0      March     Sunday  
9512    ZW                  ZWE   14439018.0      March   Saturday  

[9513 rows x 12 columns]

### Q) Filter rows where the countriesAndTerritories column contains 'Germany','France','Italy'.

In [33]:
df[df['countriesAndTerritories'].isin(['Germany','France','Italy'])]

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
3119 2020-04-08    8      4  2020   3777    1417                  France   
3120 2020-04-07    7      4  2020   3912     833                  France   
3121 2020-04-06    6      4  2020   1873     518                  France   
3122 2020-04-05    5      4  2020   4267    1053                  France   
3123 2020-04-04    4      4  2020   5233    2004                  France   
...         ...  ...    ...   ...    ...     ...                     ...   
4699 2020-01-04    4      1  2020      0       0                   Italy   
4700 2020-01-03    3      1  2020      0       0                   Italy   
4701 2020-01-02    2      1  2020      0       0                   Italy   
4702 2020-01-01    1      1  2020      0       0                   Italy   
4703 2019-12-31   31     12  2019      0       0                   Italy   

     geoId countryterritoryCode  popData2018 month_name   day_name  
3119    FR                  FRA   66987244.0      April  Wednesday  
3120    FR                  FRA   66987244.0      April    Tuesday  
3121    FR                  FRA   66987244.0      April     Monday  
3122    FR                  FRA   66987244.0      April     Sunday  
3123    FR                  FRA   66987244.0      April   Saturday  
...    ...                  ...          ...        ...        ...  
4699    IT                  ITA   60431283.0    January   Saturday  
4700    IT                  ITA   60431283.0    January     Friday  
4701    IT                  ITA   60431283.0    January   Thursday  
4702    IT                  ITA   60431283.0    January  Wednesday  
4703    IT                  ITA   60431283.0   December    Tuesday  

[300 rows x 12 columns]

### Q) Filter out rows where the countriesAndTerritories column not contains 'Germany','France','Italy'.

In [34]:
excluded_country=['Germany','France','Italy']
rest_country=df[~df['countriesAndTerritories'].isin(['excluded_country'])]
rest_country

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0    2020-04-08    8      4  2020     30       4             Afghanistan   
1    2020-04-07    7      4  2020     38       0             Afghanistan   
2    2020-04-06    6      4  2020     29       2             Afghanistan   
3    2020-04-05    5      4  2020     35       1             Afghanistan   
4    2020-04-04    4      4  2020      0       0             Afghanistan   
...         ...  ...    ...   ...    ...     ...                     ...   
9508 2020-03-25   25      3  2020      0       0                Zimbabwe   
9509 2020-03-24   24      3  2020      0       1                Zimbabwe   
9510 2020-03-23   23      3  2020      0       0                Zimbabwe   
9511 2020-03-22   22      3  2020      1       0                Zimbabwe   
9512 2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018 month_name   day_name  
0       AF                  AFG   37172386.0      April  Wednesday  
1       AF                  AFG   37172386.0      April    Tuesday  
2       AF                  AFG   37172386.0      April     Monday  
3       AF                  AFG   37172386.0      April     Sunday  
4       AF                  AFG   37172386.0      April   Saturday  
...    ...                  ...          ...        ...        ...  
9508    ZW                  ZWE   14439018.0      March  Wednesday  
9509    ZW                  ZWE   14439018.0      March    Tuesday  
9510    ZW                  ZWE   14439018.0      March     Monday  
9511    ZW                  ZWE   14439018.0      March     Sunday  
9512    ZW                  ZWE   14439018.0      March   Saturday  

[9513 rows x 12 columns]

In [35]:
df1=df[df['countriesAndTerritories'].isin(['Germany'])]

In [36]:
df1

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
3385 2020-04-08    8      4  2020   4003     254                 Germany   
3386 2020-04-07    7      4  2020   3834     173                 Germany   
3387 2020-04-06    6      4  2020   3677      92                 Germany   
3388 2020-04-05    5      4  2020   5936     184                 Germany   
3389 2020-04-04    4      4  2020   6082     141                 Germany   
...         ...  ...    ...   ...    ...     ...                     ...   
3480 2020-01-04    4      1  2020      0       0                 Germany   
3481 2020-01-03    3      1  2020      0       0                 Germany   
3482 2020-01-02    2      1  2020      0       0                 Germany   
3483 2020-01-01    1      1  2020      0       0                 Germany   
3484 2019-12-31   31     12  2019      0       0                 Germany   

     geoId countryterritoryCode  popData2018 month_name   day_name  
3385    DE                  DEU   82927922.0      April  Wednesday  
3386    DE                  DEU   82927922.0      April    Tuesday  
3387    DE                  DEU   82927922.0      April     Monday  
3388    DE                  DEU   82927922.0      April     Sunday  
3389    DE                  DEU   82927922.0      April   Saturday  
...    ...                  ...          ...        ...        ...  
3480    DE                  DEU   82927922.0    January   Saturday  
3481    DE                  DEU   82927922.0    January     Friday  
3482    DE                  DEU   82927922.0    January   Thursday  
3483    DE                  DEU   82927922.0    January  Wednesday  
3484    DE                  DEU   82927922.0   December    Tuesday  

[100 rows x 12 columns]

# reset_index()


In [37]:
df1.reset_index(drop=True,inplace=True)
# "reset_index()" used to reset the index in sequence order like 0,1,2, ....
# in this way automatically a column with index is existed. it doesnot need, so, it will remove using "drop=True"

In [38]:
df1

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0  2020-04-08    8      4  2020   4003     254                 Germany    DE   
1  2020-04-07    7      4  2020   3834     173                 Germany    DE   
2  2020-04-06    6      4  2020   3677      92                 Germany    DE   
3  2020-04-05    5      4  2020   5936     184                 Germany    DE   
4  2020-04-04    4      4  2020   6082     141                 Germany    DE   
..        ...  ...    ...   ...    ...     ...                     ...   ...   
95 2020-01-04    4      1  2020      0       0                 Germany    DE   
96 2020-01-03    3      1  2020      0       0                 Germany    DE   
97 2020-01-02    2      1  2020      0       0                 Germany    DE   
98 2020-01-01    1      1  2020      0       0                 Germany    DE   
99 2019-12-31   31     12  2019      0       0                 Germany    DE   

   countryterritoryCode  popData2018 month_name   day_name  
0                   DEU   82927922.0      April  Wednesday  
1                   DEU   82927922.0      April    Tuesday  
2                   DEU   82927922.0      April     Monday  
3                   DEU   82927922.0      April     Sunday  
4                   DEU   82927922.0      April   Saturday  
..                  ...          ...        ...        ...  
95                  DEU   82927922.0    January   Saturday  
96                  DEU   82927922.0    January     Friday  
97                  DEU   82927922.0    January   Thursday  
98                  DEU   82927922.0    January  Wednesday  
99                  DEU   82927922.0   December    Tuesday  

[100 rows x 12 columns]

### Q) extract countries and territories china

In [39]:
df_china=df[df['countriesAndTerritories'].isin(['China'])]

In [40]:
df_china

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
1845 2020-04-08    8      4  2020     86       2                   China   
1846 2020-04-07    7      4  2020     56       0                   China   
1847 2020-04-06    6      4  2020     67       2                   China   
1848 2020-04-05    5      4  2020     48       3                   China   
1849 2020-04-04    4      4  2020     62       4                   China   
...         ...  ...    ...   ...    ...     ...                     ...   
1940 2020-01-04    4      1  2020      0       0                   China   
1941 2020-01-03    3      1  2020     17       0                   China   
1942 2020-01-02    2      1  2020      0       0                   China   
1943 2020-01-01    1      1  2020      0       0                   China   
1944 2019-12-31   31     12  2019     27       0                   China   

     geoId countryterritoryCode   popData2018 month_name   day_name  
1845    CN                  CHN  1.392730e+09      April  Wednesday  
1846    CN                  CHN  1.392730e+09      April    Tuesday  
1847    CN                  CHN  1.392730e+09      April     Monday  
1848    CN                  CHN  1.392730e+09      April     Sunday  
1849    CN                  CHN  1.392730e+09      April   Saturday  
...    ...                  ...           ...        ...        ...  
1940    CN                  CHN  1.392730e+09    January   Saturday  
1941    CN                  CHN  1.392730e+09    January     Friday  
1942    CN                  CHN  1.392730e+09    January   Thursday  
1943    CN                  CHN  1.392730e+09    January  Wednesday  
1944    CN                  CHN  1.392730e+09   December    Tuesday  

[100 rows x 12 columns]

In [41]:
df_china.reset_index(drop=True,inplace=True)

In [42]:
df_china

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0  2020-04-08    8      4  2020     86       2                   China    CN   
1  2020-04-07    7      4  2020     56       0                   China    CN   
2  2020-04-06    6      4  2020     67       2                   China    CN   
3  2020-04-05    5      4  2020     48       3                   China    CN   
4  2020-04-04    4      4  2020     62       4                   China    CN   
..        ...  ...    ...   ...    ...     ...                     ...   ...   
95 2020-01-04    4      1  2020      0       0                   China    CN   
96 2020-01-03    3      1  2020     17       0                   China    CN   
97 2020-01-02    2      1  2020      0       0                   China    CN   
98 2020-01-01    1      1  2020      0       0                   China    CN   
99 2019-12-31   31     12  2019     27       0                   China    CN   

   countryterritoryCode   popData2018 month_name   day_name  
0                   CHN  1.392730e+09      April  Wednesday  
1                   CHN  1.392730e+09      April    Tuesday  
2                   CHN  1.392730e+09      April     Monday  
3                   CHN  1.392730e+09      April     Sunday  
4                   CHN  1.392730e+09      April   Saturday  
..                  ...           ...        ...        ...  
95                  CHN  1.392730e+09    January   Saturday  
96                  CHN  1.392730e+09    January     Friday  
97                  CHN  1.392730e+09    January   Thursday  
98                  CHN  1.392730e+09    January  Wednesday  
99                  CHN  1.392730e+09   December    Tuesday  

[100 rows x 12 columns]

# customizing_ index
set_index()

In [43]:
df.set_index(['countriesAndTerritories'])

dateRep  day  month  year  cases  deaths geoId  \
countriesAndTerritories                                                     
Afghanistan             2020-04-08    8      4  2020     30       4    AF   
Afghanistan             2020-04-07    7      4  2020     38       0    AF   
Afghanistan             2020-04-06    6      4  2020     29       2    AF   
Afghanistan             2020-04-05    5      4  2020     35       1    AF   
Afghanistan             2020-04-04    4      4  2020      0       0    AF   
...                            ...  ...    ...   ...    ...     ...   ...   
Zimbabwe                2020-03-25   25      3  2020      0       0    ZW   
Zimbabwe                2020-03-24   24      3  2020      0       1    ZW   
Zimbabwe                2020-03-23   23      3  2020      0       0    ZW   
Zimbabwe                2020-03-22   22      3  2020      1       0    ZW   
Zimbabwe                2020-03-21   21      3  2020      1       0    ZW   

                        countryterritoryCode  popData2018 month_name  \
countriesAndTerritories                                                
Afghanistan                              AFG   37172386.0      April   
Afghanistan                              AFG   37172386.0      April   
Afghanistan                              AFG   37172386.0      April   
Afghanistan                              AFG   37172386.0      April   
Afghanistan                              AFG   37172386.0      April   
...                                      ...          ...        ...   
Zimbabwe                                 ZWE   14439018.0      March   
Zimbabwe                                 ZWE   14439018.0      March   
Zimbabwe                                 ZWE   14439018.0      March   
Zimbabwe                                 ZWE   14439018.0      March   
Zimbabwe                                 ZWE   14439018.0      March   

                          day_name  
countriesAndTerritories             
Afghanistan              Wednesday  
Afghanistan                Tuesday  
Afghanistan                 Monday  
Afghanistan                 Sunday  
Afghanistan               Saturday  
...                            ...  
Zimbabwe                 Wednesday  
Zimbabwe                   Tuesday  
Zimbabwe                    Monday  
Zimbabwe                    Sunday  
Zimbabwe                  Saturday  

[9513 rows x 11 columns]

###  Q) extract geoid HR,CU,CW,CY -- reindex,year wise indexing

In [44]:
df_geoid=df[df['geoId'].isin(['HR','CU','CW','CY'])]

In [45]:
df_geoid

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
2056 2020-04-08    8      4  2020     60       2                 Croatia   
2057 2020-04-07    7      4  2020     40       1                 Croatia   
2058 2020-04-06    6      4  2020     56       3                 Croatia   
2059 2020-04-05    5      4  2020     47       4                 Croatia   
2060 2020-04-04    4      4  2020     68       1                 Croatia   
...         ...  ...    ...   ...    ...     ...                     ...   
2217 2020-03-16   16      3  2020      9       0                  Cyprus   
2218 2020-03-15   15      3  2020      7       0                  Cyprus   
2219 2020-03-14   14      3  2020      8       0                  Cyprus   
2220 2020-03-12   12      3  2020      4       0                  Cyprus   
2221 2020-03-10   10      3  2020      2       0                  Cyprus   

     geoId countryterritoryCode  popData2018 month_name   day_name  
2056    HR                  HRV    4089400.0      April  Wednesday  
2057    HR                  HRV    4089400.0      April    Tuesday  
2058    HR                  HRV    4089400.0      April     Monday  
2059    HR                  HRV    4089400.0      April     Sunday  
2060    HR                  HRV    4089400.0      April   Saturday  
...    ...                  ...          ...        ...        ...  
2217    CY                  CYP    1189265.0      March     Monday  
2218    CY                  CYP    1189265.0      March     Sunday  
2219    CY                  CYP    1189265.0      March   Saturday  
2220    CY                  CYP    1189265.0      March   Thursday  
2221    CY                  CYP    1189265.0      March    Tuesday  

[166 rows x 12 columns]

In [46]:
df_geoid.reset_index(drop=True,inplace=True)

In [47]:
df_geoid

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0   2020-04-08    8      4  2020     60       2                 Croatia    HR   
1   2020-04-07    7      4  2020     40       1                 Croatia    HR   
2   2020-04-06    6      4  2020     56       3                 Croatia    HR   
3   2020-04-05    5      4  2020     47       4                 Croatia    HR   
4   2020-04-04    4      4  2020     68       1                 Croatia    HR   
..         ...  ...    ...   ...    ...     ...                     ...   ...   
161 2020-03-16   16      3  2020      9       0                  Cyprus    CY   
162 2020-03-15   15      3  2020      7       0                  Cyprus    CY   
163 2020-03-14   14      3  2020      8       0                  Cyprus    CY   
164 2020-03-12   12      3  2020      4       0                  Cyprus    CY   
165 2020-03-10   10      3  2020      2       0                  Cyprus    CY   

    countryterritoryCode  popData2018 month_name   day_name  
0                    HRV    4089400.0      April  Wednesday  
1                    HRV    4089400.0      April    Tuesday  
2                    HRV    4089400.0      April     Monday  
3                    HRV    4089400.0      April     Sunday  
4                    HRV    4089400.0      April   Saturday  
..                   ...          ...        ...        ...  
161                  CYP    1189265.0      March     Monday  
162                  CYP    1189265.0      March     Sunday  
163                  CYP    1189265.0      March   Saturday  
164                  CYP    1189265.0      March   Thursday  
165                  CYP    1189265.0      March    Tuesday  

[166 rows x 12 columns]

In [48]:
df_geoid=df_geoid.set_index(['month_name'])

In [49]:
df_geoid

dateRep  day  month  year  cases  deaths  \
month_name                                               
April      2020-04-08    8      4  2020     60       2   
April      2020-04-07    7      4  2020     40       1   
April      2020-04-06    6      4  2020     56       3   
April      2020-04-05    5      4  2020     47       4   
April      2020-04-04    4      4  2020     68       1   
...               ...  ...    ...   ...    ...     ...   
March      2020-03-16   16      3  2020      9       0   
March      2020-03-15   15      3  2020      7       0   
March      2020-03-14   14      3  2020      8       0   
March      2020-03-12   12      3  2020      4       0   
March      2020-03-10   10      3  2020      2       0   

           countriesAndTerritories geoId countryterritoryCode  popData2018  \
month_name                                                                   
April                      Croatia    HR                  HRV    4089400.0   
April                      Croatia    HR                  HRV    4089400.0   
April                      Croatia    HR                  HRV    4089400.0   
April                      Croatia    HR                  HRV    4089400.0   
April                      Croatia    HR                  HRV    4089400.0   
...                            ...   ...                  ...          ...   
March                       Cyprus    CY                  CYP    1189265.0   
March                       Cyprus    CY                  CYP    1189265.0   
March                       Cyprus    CY                  CYP    1189265.0   
March                       Cyprus    CY                  CYP    1189265.0   
March                       Cyprus    CY                  CYP    1189265.0   

             day_name  
month_name             
April       Wednesday  
April         Tuesday  
April          Monday  
April          Sunday  
April        Saturday  
...               ...  
March          Monday  
March          Sunday  
March        Saturday  
March        Thursday  
March         Tuesday  

[166 rows x 11 columns]

### Q) filter the DataFrame to include only the records from the countries with the territory codes 'GVR', 'KNA', 'NLD', and 'PRT', then set the index to 'day_name', and finally sort the DataFrame by the 'deaths' column in descending order

In [50]:
df_dayname=df[df['countryterritoryCode'].isin(['GVR','KNA','NLD','PRT'])].set_index(['day_name']).sort_values(by='deaths',ascending=False)

In [51]:
df_dayname

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
day_name                                                                        
Wednesday 2020-04-08    8      4  2020    777     234             Netherlands   
Wednesday 2020-04-01    1      4  2020    845     175             Netherlands   
Friday    2020-04-03    3      4  2020   1083     166             Netherlands   
Sunday    2020-04-05    5      4  2020    904     164             Netherlands   
Saturday  2020-04-04    4      4  2020   1026     148             Netherlands   
...              ...  ...    ...   ...    ...     ...                     ...   
Wednesday 2020-02-05    5      2  2020      0       0             Netherlands   
Thursday  2020-02-06    6      2  2020      0       0             Netherlands   
Friday    2020-02-07    7      2  2020      0       0             Netherlands   
Saturday  2020-02-08    8      2  2020      0       0             Netherlands   
Thursday  2020-03-26   26      3  2020      2       0   Saint_Kitts_and_Nevis   

          geoId countryterritoryCode  popData2018 month_name  
day_name                                                      
Wednesday    NL                  NLD   17231017.0      April  
Wednesday    NL                  NLD   17231017.0      April  
Friday       NL                  NLD   17231017.0      April  
Sunday       NL                  NLD   17231017.0      April  
Saturday     NL                  NLD   17231017.0      April  
...         ...                  ...          ...        ...  
Wednesday    NL                  NLD   17231017.0   February  
Thursday     NL                  NLD   17231017.0   February  
Friday       NL                  NLD   17231017.0   February  
Saturday     NL                  NLD   17231017.0   February  
Thursday     KN                  KNA      52441.0      March  

[151 rows x 11 columns]

###  Q) filter the rows where cases>1000 and deaths>500

In [52]:
deaths_cases=df[(df['cases']>1000)&(df['deaths']>500)]

In [53]:
deaths_cases

dateRep  day  month  year  cases  deaths   countriesAndTerritories  \
3119 2020-04-08    8      4  2020   3777    1417                    France   
3120 2020-04-07    7      4  2020   3912     833                    France   
3121 2020-04-06    6      4  2020   1873     518                    France   
3122 2020-04-05    5      4  2020   4267    1053                    France   
3123 2020-04-04    4      4  2020   5233    2004                    France   
3125 2020-04-02    2      4  2020   4861     509                    France   
4604 2020-04-08    8      4  2020   3039     604                     Italy   
4605 2020-04-07    7      4  2020   3599     636                     Italy   
4606 2020-04-06    6      4  2020   4316     527                     Italy   
4607 2020-04-05    5      4  2020   4805     681                     Italy   
4608 2020-04-04    4      4  2020   4585     764                     Italy   
4609 2020-04-03    3      4  2020   4668     760                     Italy   
4610 2020-04-02    2      4  2020   4782     727                     Italy   
4611 2020-04-01    1      4  2020   4053     839                     Italy   
4612 2020-03-31   31      3  2020   4050     810                     Italy   
4613 2020-03-30   30      3  2020   5217     758                     Italy   
4614 2020-03-29   29      3  2020   5974     887                     Italy   
4615 2020-03-28   28      3  2020   5959     971                     Italy   
4616 2020-03-27   27      3  2020   6153     660                     Italy   
4617 2020-03-26   26      3  2020   5210     685                     Italy   
4618 2020-03-25   25      3  2020   5249     743                     Italy   
4619 2020-03-24   24      3  2020   4789     601                     Italy   
4620 2020-03-23   23      3  2020   5560     649                     Italy   
4621 2020-03-22   22      3  2020   6557     795                     Italy   
4622 2020-03-21   21      3  2020   5986     625                     Italy   
8139 2020-04-08    8      4  2020   5478     743                     Spain   
8140 2020-04-07    7      4  2020   4273     637                     Spain   
8141 2020-04-06    6      4  2020   6023     674                     Spain   
8142 2020-04-05    5      4  2020   7026     809                     Spain   
8143 2020-04-04    4      4  2020   7472     932                     Spain   
8144 2020-04-03    3      4  2020   8102     950                     Spain   
8145 2020-04-02    2      4  2020   7719     864                     Spain   
8146 2020-04-01    1      4  2020   9222     849                     Spain   
8147 2020-03-31   31      3  2020   6398     812                     Spain   
8148 2020-03-30   30      3  2020   6549     838                     Spain   
8149 2020-03-29   29      3  2020   8189     832                     Spain   
8150 2020-03-28   28      3  2020   7871     769                     Spain   
8151 2020-03-27   27      3  2020   8578     655                     Spain   
8152 2020-03-26   26      3  2020   7937     738                     Spain   
8153 2020-03-25   25      3  2020   6584     514                     Spain   
9065 2020-04-08    8      4  2020   3634     786            United_Kingdom   
9067 2020-04-06    6      4  2020   5903     621            United_Kingdom   
9068 2020-04-05    5      4  2020   3735     708            United_Kingdom   
9069 2020-04-04    4      4  2020   4450     684            United_Kingdom   
9071 2020-04-02    2      4  2020   4324     743            United_Kingdom   
9188 2020-04-08    8      4  2020  30613    1906  United_States_of_America   
9189 2020-04-07    7      4  2020  30561    1342  United_States_of_America   
9190 2020-04-06    6      4  2020  25398    1146  United_States_of_America   
9191 2020-04-05    5      4  2020  34272    1344  United_States_of_America   
9192 2020-04-04    4      4  2020  32425    1104  United_States_of_America   
9193 2020-04-03    3      4  2

In [54]:
# Q) reset index of deaths_cases
deaths_cases.reset_index(drop=True,inplace=True)

In [55]:
deaths_cases

dateRep  day  month  year  cases  deaths   countriesAndTerritories  \
0  2020-04-08    8      4  2020   3777    1417                    France   
1  2020-04-07    7      4  2020   3912     833                    France   
2  2020-04-06    6      4  2020   1873     518                    France   
3  2020-04-05    5      4  2020   4267    1053                    France   
4  2020-04-04    4      4  2020   5233    2004                    France   
5  2020-04-02    2      4  2020   4861     509                    France   
6  2020-04-08    8      4  2020   3039     604                     Italy   
7  2020-04-07    7      4  2020   3599     636                     Italy   
8  2020-04-06    6      4  2020   4316     527                     Italy   
9  2020-04-05    5      4  2020   4805     681                     Italy   
10 2020-04-04    4      4  2020   4585     764                     Italy   
11 2020-04-03    3      4  2020   4668     760                     Italy   
12 2020-04-02    2      4  2020   4782     727                     Italy   
13 2020-04-01    1      4  2020   4053     839                     Italy   
14 2020-03-31   31      3  2020   4050     810                     Italy   
15 2020-03-30   30      3  2020   5217     758                     Italy   
16 2020-03-29   29      3  2020   5974     887                     Italy   
17 2020-03-28   28      3  2020   5959     971                     Italy   
18 2020-03-27   27      3  2020   6153     660                     Italy   
19 2020-03-26   26      3  2020   5210     685                     Italy   
20 2020-03-25   25      3  2020   5249     743                     Italy   
21 2020-03-24   24      3  2020   4789     601                     Italy   
22 2020-03-23   23      3  2020   5560     649                     Italy   
23 2020-03-22   22      3  2020   6557     795                     Italy   
24 2020-03-21   21      3  2020   5986     625                     Italy   
25 2020-04-08    8      4  2020   5478     743                     Spain   
26 2020-04-07    7      4  2020   4273     637                     Spain   
27 2020-04-06    6      4  2020   6023     674                     Spain   
28 2020-04-05    5      4  2020   7026     809                     Spain   
29 2020-04-04    4      4  2020   7472     932                     Spain   
30 2020-04-03    3      4  2020   8102     950                     Spain   
31 2020-04-02    2      4  2020   7719     864                     Spain   
32 2020-04-01    1      4  2020   9222     849                     Spain   
33 2020-03-31   31      3  2020   6398     812                     Spain   
34 2020-03-30   30      3  2020   6549     838                     Spain   
35 2020-03-29   29      3  2020   8189     832                     Spain   
36 2020-03-28   28      3  2020   7871     769                     Spain   
37 2020-03-27   27      3  2020   8578     655                     Spain   
38 2020-03-26   26      3  2020   7937     738                     Spain   
39 2020-03-25   25      3  2020   6584     514                     Spain   
40 2020-04-08    8      4  2020   3634     786            United_Kingdom   
41 2020-04-06    6      4  2020   5903     621            United_Kingdom   
42 2020-04-05    5      4  2020   3735     708            United_Kingdom   
43 2020-04-04    4      4  2020   4450     684            United_Kingdom   
44 2020-04-02    2      4  2020   4324     743            United_Kingdom   
45 2020-04-08    8      4  2020  30613    1906  United_States_of_America   
46 2020-04-07    7      4  2020  30561    1342  United_States_of_America   
47 2020-04-06    6      4  2020  25398    1146  United_States_of_America   
48 2020-04-05    5      4  2020  34272    1344  United_States_of_America   
49 2020-04-04    4      4  2020  32425    1104  United_States_of_America   
50 2020-04-03    3      4  2020  28819     915  United_States_of_America   
51 2020-04-02    2      4  2020  27103    1059  United

### Q) set_index is month_name and sort in descending

In [56]:
deaths_cases.set_index('month_name').sort_values(by='deaths',ascending=False)

dateRep  day  month  year  cases  deaths  \
month_name                                               
April      2020-04-04    4      4  2020   5233    2004   
April      2020-04-08    8      4  2020  30613    1906   
April      2020-04-08    8      4  2020   3777    1417   
April      2020-04-05    5      4  2020  34272    1344   
April      2020-04-07    7      4  2020  30561    1342   
April      2020-04-06    6      4  2020  25398    1146   
April      2020-04-04    4      4  2020  32425    1104   
April      2020-04-02    2      4  2020  27103    1059   
April      2020-04-05    5      4  2020   4267    1053   
March      2020-03-28   28      3  2020   5959     971   
April      2020-04-03    3      4  2020   8102     950   
April      2020-04-04    4      4  2020   7472     932   
April      2020-04-03    3      4  2020  28819     915   
April      2020-04-01    1      4  2020  24998     909   
March      2020-03-29   29      3  2020   5974     887   
April      2020-04-02    2      4  2020   7719     864   
April      2020-04-01    1      4  2020   9222     849   
April      2020-04-01    1      4  2020   4053     839   
March      2020-03-30   30      3  2020   6549     838   
April      2020-04-07    7      4  2020   3912     833   
March      2020-03-29   29      3  2020   8189     832   
March      2020-03-31   31      3  2020   6398     812   
March      2020-03-31   31      3  2020   4050     810   
April      2020-04-05    5      4  2020   7026     809   
March      2020-03-22   22      3  2020   6557     795   
April      2020-04-08    8      4  2020   3634     786   
March      2020-03-28   28      3  2020   7871     769   
April      2020-04-04    4      4  2020   4585     764   
April      2020-04-03    3      4  2020   4668     760   
March      2020-03-30   30      3  2020   5217     758   
April      2020-04-08    8      4  2020   5478     743   
March      2020-03-25   25      3  2020   5249     743   
April      2020-04-02    2      4  2020   4324     743   
March      2020-03-26   26      3  2020   7937     738   
April      2020-04-02    2      4  2020   4782     727   
April      2020-04-05    5      4  2020   3735     708   
March      2020-03-26   26      3  2020   5210     685   
April      2020-04-04    4      4  2020   4450     684   
April      2020-04-05    5      4  2020   4805     681   
April      2020-04-06    6      4  2020   6023     674   
March      2020-03-31   31      3  2020  21595     661   
March      2020-03-27   27      3  2020   6153     660   
March      2020-03-27   27      3  2020   8578     655   
March      2020-03-23   23      3  2020   5560     649   
April      2020-04-07    7      4  2020   4273     637   
April      2020-04-07    7      4  2020   3599     636   
March      2020-03-21   21      3  2020   5986     625   
April      2020-04-06    6      4  2020   5903     621   
April      2020-04-08    8      4  2020   3039     604   
March      2020-03-24   24      3  2020   4789     601   
April      2020-04-06    6      4  2020   4316     527   
April      2020-04-06    6      4  2020   1873     518   
March      2020-03-25   25      3  2020   6584     514   
April      2020-04-02    2      4  2020   4861     509   

             countriesAndTerritories geoId countryterritoryCode  popData2018  \
month_name                                                                     
April                         France    FR                  FRA   66987244.0   
April       United_States_of_America    US                  USA  327167434.0   
April                         France    FR                  FRA   66987244.0   
April       United_States_of_America    US                  USA  327167434.0   
April       United_States_of_America    US                  USA  327167434.0   
April       United_States_of_America    US                  USA  327167434.0   
April       United_States_of_America    US                  USA  327167434.0   
April       United_States_of_America    US   

## Pivot_Table()

*   The pivot function in Pandas is used to reshape or transform data by pivoting the values of one column into multiple columns.
*   pivot is used to transform long-form data to wide-form

### Q) create a pivot table that summarizes the total number of deaths for each country, broken down by month

In [57]:
df.pivot_table(index='countriesAndTerritories',columns="month_name",values='deaths',aggfunc='sum')

month_name               April  December  February  January  March
countriesAndTerritories                                           
Afghanistan                7.0       0.0       0.0      0.0    4.0
Albania                   10.0       NaN       NaN      NaN   12.0
Algeria                  163.0       0.0       0.0      0.0   31.0
Andorra                   14.0       NaN       NaN      NaN    8.0
Angola                     0.0       NaN       NaN      NaN    2.0
...                        ...       ...       ...      ...    ...
Uzbekistan                 0.0       NaN       NaN      NaN    2.0
Venezuela                  4.0       NaN       NaN      NaN    3.0
Vietnam                    0.0       0.0       0.0      0.0    0.0
Zambia                     1.0       NaN       NaN      NaN    0.0
Zimbabwe                   0.0       NaN       NaN      NaN    1.0

[204 rows x 5 columns]

### Q) create a pivot table that shows the average number of deaths per country for each month

In [58]:
df.pivot_table(index='countriesAndTerritories',columns="month_name",values='deaths',aggfunc='mean')

month_name                April  December  February  January     March
countriesAndTerritories                                               
Afghanistan               0.875       0.0       0.0      0.0  0.190476
Albania                   1.250       NaN       NaN      NaN  0.521739
Algeria                  20.375       0.0       0.0      0.0  1.192308
Andorra                   1.750       NaN       NaN      NaN  0.444444
Angola                    0.000       NaN       NaN      NaN  0.200000
...                         ...       ...       ...      ...       ...
Uzbekistan                0.000       NaN       NaN      NaN  0.125000
Venezuela                 0.500       NaN       NaN      NaN  0.176471
Vietnam                   0.000       0.0       0.0      0.0  0.000000
Zambia                    0.125       NaN       NaN      NaN  0.000000
Zimbabwe                  0.000       NaN       NaN      NaN  0.090909

[204 rows x 5 columns]

### Q) create a pivot table that summarizes the total number of deaths for each country, broken down by year

In [59]:
df.pivot_table(index='countriesAndTerritories',columns="year",values='deaths',aggfunc='sum')

year                     2019   2020
countriesAndTerritories             
Afghanistan               0.0   11.0
Albania                   NaN   22.0
Algeria                   0.0  194.0
Andorra                   NaN   22.0
Angola                    NaN    2.0
...                       ...    ...
Uzbekistan                NaN    2.0
Venezuela                 NaN    7.0
Vietnam                   0.0    0.0
Zambia                    NaN    1.0
Zimbabwe                  NaN    1.0

[204 rows x 2 columns]

### Q) create a pivot table that displays the average population data for each country across different years

In [60]:
df.pivot_table(index='year',columns='countriesAndTerritories',values='popData2018',aggfunc='mean')

countriesAndTerritories  Afghanistan    Albania     Algeria  Andorra  \
year                                                                   
2019                      37172386.0        NaN  42228429.0      NaN   
2020                      37172386.0  2866376.0  42228429.0  77006.0   

countriesAndTerritories      Angola  Antigua_and_Barbuda   Argentina  \
year                                                                   
2019                            NaN                  NaN         NaN   
2020                     30809762.0              96286.0  44494502.0   

countriesAndTerritories    Armenia     Aruba   Australia  ...  United_Kingdom  \
year                                                      ...                   
2019                     2951776.0       NaN  24992369.0  ...      66488991.0   
2020                     2951776.0  105845.0  24992369.0  ...      66488991.0   

countriesAndTerritories  United_Republic_of_Tanzania  \
year                                                   
2019                                             NaN   
2020                                      56318348.0   

countriesAndTerritories  United_States_Virgin_Islands  \
year                                                    
2019                                              NaN   
2020                                         106977.0   

countriesAndTerritories  United_States_of_America    Uruguay  Uzbekistan  \
year                                                                       
2019                                  327167434.0        NaN         NaN   
2020                                  327167434.0  3449299.0  32955400.0   

countriesAndTerritories   Venezuela     Vietnam      Zambia    Zimbabwe  
year                                                                     
2019                            NaN  95540395.0         NaN         NaN  
2020                     28870195.0  95540395.0  17351822.0  14439018.0  

[2 rows x 199 columns]

### Q) create a pivot table that shows the maximum number of cases recorded for each country in each year

In [61]:
df.pivot_table(index='countriesAndTerritories',columns="year",values='cases',aggfunc='max')

year                     2019   2020
countriesAndTerritories             
Afghanistan               0.0   43.0
Albania                   NaN   29.0
Algeria                   0.0  314.0
Andorra                   NaN   43.0
Angola                    NaN    4.0
...                       ...    ...
Uzbekistan                NaN  107.0
Venezuela                 NaN   48.0
Vietnam                   0.0   54.0
Zambia                    NaN   12.0
Zimbabwe                  NaN    2.0

[204 rows x 2 columns]

### Q)  create a pivot table that displays the maximum number of cases recorded for each country during each month

In [62]:
df.pivot_table(index='countriesAndTerritories',columns='month_name',values='cases',aggfunc='max')

month_name               April  December  February  January  March
countriesAndTerritories                                           
Afghanistan               43.0       0.0       1.0      0.0   33.0
Albania                   29.0       NaN       NaN      NaN   28.0
Algeria                  314.0       0.0       1.0      0.0  104.0
Andorra                   38.0       NaN       NaN      NaN   43.0
Angola                     4.0       NaN       NaN      NaN    3.0
...                        ...       ...       ...      ...    ...
Uzbekistan               107.0       NaN       NaN      NaN   29.0
Venezuela                 11.0       NaN       NaN      NaN   48.0
Vietnam                    6.0       0.0       2.0      3.0   54.0
Zambia                     3.0       NaN       NaN      NaN   12.0
Zimbabwe                   1.0       NaN       NaN      NaN    2.0

[204 rows x 5 columns]

### Q) create a pivot table that shows the total number of deaths for each geographical region, broken down by each day of the week

In [63]:
pt=df.pivot_table(index='geoId',columns='day_name',values='deaths',aggfunc='sum')

In [64]:
pt

day_name  Friday  Monday  Saturday  Sunday  Thursday  Tuesday  Wednesday
geoId                                                                   
AD           4.0     3.0       1.0     2.0       2.0      5.0        5.0
AE           2.0     0.0       0.0     4.0       0.0      4.0        2.0
AF           0.0     3.0       1.0     2.0       0.0      1.0        4.0
AG           0.0     0.0       0.0     0.0       0.0      0.0        0.0
AI           0.0     0.0       0.0     0.0       0.0      0.0        0.0
...          ...     ...       ...     ...       ...      ...        ...
VN           0.0     0.0       0.0     0.0       0.0      0.0        0.0
XK           0.0     1.0       0.0     0.0       0.0      2.0        2.0
ZA           2.0     2.0       2.0     2.0       0.0      2.0        3.0
ZM           1.0     0.0       0.0     0.0       0.0      0.0        0.0
ZW           0.0     0.0       0.0     0.0       0.0      1.0        0.0

[203 rows x 7 columns]

## melt()

* it’s a tool for reshaping data, turning columns into rows, thus ‘melting’ the data structure.

In [65]:
df.columns

Index(['dateRep', 'day', 'month', 'year', 'cases', 'deaths',
       'countriesAndTerritories', 'geoId', 'countryterritoryCode',
       'popData2018', 'month_name', 'day_name'],
      dtype='object')

### Q) reshape a DataFrame to unpivot the 'cases' and 'deaths' columns into a single column, while keeping the countries and territories and reported date information intact

In [66]:
death_cases=df.melt(id_vars=['countriesAndTerritories','dateRep'],value_vars=['cases','deaths'],var_name='death_cases',value_name='count')
death_cases

countriesAndTerritories    dateRep death_cases  count
0                 Afghanistan 2020-04-08       cases     30
1                 Afghanistan 2020-04-07       cases     38
2                 Afghanistan 2020-04-06       cases     29
3                 Afghanistan 2020-04-05       cases     35
4                 Afghanistan 2020-04-04       cases      0
...                       ...        ...         ...    ...
19021                Zimbabwe 2020-03-25      deaths      0
19022                Zimbabwe 2020-03-24      deaths      1
19023                Zimbabwe 2020-03-23      deaths      0
19024                Zimbabwe 2020-03-22      deaths      0
19025                Zimbabwe 2020-03-21      deaths      0

[19026 rows x 4 columns]

In [67]:
death_cases['death_cases'].value_counts()

death_cases
cases     9513
deaths    9513
Name: count, dtype: int64

In [68]:
# method 2
death_cases.value_counts('death_cases')

death_cases
cases     9513
deaths    9513
Name: count, dtype: int64

In [69]:
df.pivot_table(index='geoId',columns='day_name',values='cases',aggfunc='max')

day_name  Friday  Monday  Saturday  Sunday  Thursday  Tuesday  Wednesday
geoId                                                                   
AD          38.0    35.0      43.0    41.0      39.0     36.0       31.0
AE         360.0   294.0      72.0   481.0      85.0    277.0      283.0
AF          43.0    29.0      16.0    35.0      33.0     38.0       30.0
AG           4.0     0.0       6.0     1.0       0.0      0.0        2.0
AI           2.0     0.0       0.0     0.0       0.0      0.0        0.0
...          ...     ...       ...     ...       ...      ...        ...
VN           9.0    24.0      16.0    54.0      15.0      5.0       11.0
XK          13.0     8.0       7.0     6.0       8.0     30.0       39.0
ZA         218.0    93.0     243.0    80.0     152.0    128.0      155.0
ZM           3.0     1.0       2.0    12.0       9.0      6.0        0.0
ZW           0.0     0.0       2.0     2.0       1.0      0.0        1.0

[203 rows x 7 columns]

## groupby()

*   The groupby function in pandas is used to split the data into groups based on some criteria, apply a function to each group independently, and then combine the results.
*   Summarized by the "split-apply-combine" pattern.

### Q)  find the maximum number of deaths recorded for each country or territory in the DataFrame

In [70]:
df.groupby('countriesAndTerritories')['deaths'].max()

countriesAndTerritories
Afghanistan     4
Albania         3
Algeria        47
Andorra         4
Angola          2
               ..
Uzbekistan      1
Venezuela       2
Vietnam         0
Zambia          1
Zimbabwe        1
Name: deaths, Length: 204, dtype: int64

### Q) calculate the average number of cases for each country or territory in the DataFrame

In [71]:
df.groupby('countriesAndTerritories')['cases'].mean()

countriesAndTerritories
Afghanistan     4.077778
Albania        12.354839
Algeria        15.452632
Andorra        20.961538
Angola          0.944444
                 ...    
Uzbekistan     21.000000
Venezuela       6.640000
Vietnam         2.614583
Zambia          1.857143
Zimbabwe        0.526316
Name: cases, Length: 204, dtype: float64

### Q) calculate the total number of deaths for each country or territory for each year in the DataFrame

In [72]:
df.groupby(['countriesAndTerritories','year'])['deaths'].sum()

countriesAndTerritories  year
Afghanistan              2019      0
                         2020     11
Albania                  2020     22
Algeria                  2019      0
                         2020    194
                                ... 
Venezuela                2020      7
Vietnam                  2019      0
                         2020      0
Zambia                   2020      1
Zimbabwe                 2020      1
Name: deaths, Length: 271, dtype: int64

### Q) find the average population data for the year 2018 for each country or territory in the DataFrame

In [73]:
df.groupby('countriesAndTerritories')['popData2018'].mean()

countriesAndTerritories
Afghanistan    37172386.0
Albania         2866376.0
Algeria        42228429.0
Andorra           77006.0
Angola         30809762.0
                  ...    
Uzbekistan     32955400.0
Venezuela      28870195.0
Vietnam        95540395.0
Zambia         17351822.0
Zimbabwe       14439018.0
Name: popData2018, Length: 204, dtype: float64

### Q)  determine the maximum number of cases and the minimum number of deaths for each country or territory in each year in the DataFrame

In [74]:
df.groupby(['countriesAndTerritories','year']).agg({'cases':max,'deaths':min})

cases  deaths
countriesAndTerritories year               
Afghanistan             2019      0       0
                        2020     43       0
Albania                 2020     29       0
Algeria                 2019      0       0
                        2020    314       0
...                             ...     ...
Venezuela               2020     48       0
Vietnam                 2019      0       0
                        2020     54       0
Zambia                  2020     12       0
Zimbabwe                2020      2       0

[271 rows x 2 columns]

### Q) find the minimum, maximum, and average population data for the year 2018 for each country or territory in each year in the DataFrame

In [75]:
df.groupby(['countriesAndTerritories','year'])['popData2018'].agg(['min','max','mean'])

min         max        mean
countriesAndTerritories year                                    
Afghanistan             2019  37172386.0  37172386.0  37172386.0
                        2020  37172386.0  37172386.0  37172386.0
Albania                 2020   2866376.0   2866376.0   2866376.0
Algeria                 2019  42228429.0  42228429.0  42228429.0
                        2020  42228429.0  42228429.0  42228429.0
...                                  ...         ...         ...
Venezuela               2020  28870195.0  28870195.0  28870195.0
Vietnam                 2019  95540395.0  95540395.0  95540395.0
                        2020  95540395.0  95540395.0  95540395.0
Zambia                  2020  17351822.0  17351822.0  17351822.0
Zimbabwe                2020  14439018.0  14439018.0  14439018.0

[271 rows x 3 columns]

In [76]:
df.columns

Index(['dateRep', 'day', 'month', 'year', 'cases', 'deaths',
       'countriesAndTerritories', 'geoId', 'countryterritoryCode',
       'popData2018', 'month_name', 'day_name'],
      dtype='object')

### Q) find the maximum population data for 2018 and the total number of cases for each combination of date, day, month, and year in the DataFrame

In [77]:
df.groupby(['dateRep','day', 'month', 'year']).agg({ 'popData2018':max,'cases':sum})

popData2018  cases
dateRep    day month year                     
2019-12-31 31  12    2019  1.392730e+09     27
2020-01-01 1   1     2020  1.392730e+09      0
2020-01-02 2   1     2020  1.392730e+09      0
2020-01-03 3   1     2020  1.392730e+09     17
2020-01-04 4   1     2020  1.392730e+09      0
...                                 ...    ...
2020-04-04 4   4     2020  1.392730e+09  81714
2020-04-05 5   4     2020  1.392730e+09  86713
2020-04-06 6   4     2020  1.392730e+09  71232
2020-04-07 7   4     2020  1.392730e+09  71392
2020-04-08 8   4     2020  1.392730e+09  74902

[100 rows x 2 columns]

## apply()
*  The apply() method is one of the most common methods of data preprocessing.
* It simplifies applying a function on each row or column in a pandas DataFrame.

### Q) create a new column in the DataFrame that adds 5 to the value in the 'popData2018' column for each row

In [78]:
def add_5(data):
    pop_sum=data['popData2018']+5
    return pop_sum


In [79]:
df['pop_sum']=df.apply(lambda x:add_5(x),axis=1)

In [80]:
df

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0    2020-04-08    8      4  2020     30       4             Afghanistan   
1    2020-04-07    7      4  2020     38       0             Afghanistan   
2    2020-04-06    6      4  2020     29       2             Afghanistan   
3    2020-04-05    5      4  2020     35       1             Afghanistan   
4    2020-04-04    4      4  2020      0       0             Afghanistan   
...         ...  ...    ...   ...    ...     ...                     ...   
9508 2020-03-25   25      3  2020      0       0                Zimbabwe   
9509 2020-03-24   24      3  2020      0       1                Zimbabwe   
9510 2020-03-23   23      3  2020      0       0                Zimbabwe   
9511 2020-03-22   22      3  2020      1       0                Zimbabwe   
9512 2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018 month_name   day_name     pop_sum  
0       AF                  AFG   37172386.0      April  Wednesday  37172391.0  
1       AF                  AFG   37172386.0      April    Tuesday  37172391.0  
2       AF                  AFG   37172386.0      April     Monday  37172391.0  
3       AF                  AFG   37172386.0      April     Sunday  37172391.0  
4       AF                  AFG   37172386.0      April   Saturday  37172391.0  
...    ...                  ...          ...        ...        ...         ...  
9508    ZW                  ZWE   14439018.0      March  Wednesday  14439023.0  
9509    ZW                  ZWE   14439018.0      March    Tuesday  14439023.0  
9510    ZW                  ZWE   14439018.0      March     Monday  14439023.0  
9511    ZW                  ZWE   14439018.0      March     Sunday  14439023.0  
9512    ZW                  ZWE   14439018.0      March   Saturday  14439023.0  

[9513 rows x 13 columns]

### Q) create a new column in the DataFrame that adds 5 to each value in the 'deaths' column

In [81]:
df['death_5']=df['deaths'].apply(lambda x:x+5)

In [82]:
df

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0    2020-04-08    8      4  2020     30       4             Afghanistan   
1    2020-04-07    7      4  2020     38       0             Afghanistan   
2    2020-04-06    6      4  2020     29       2             Afghanistan   
3    2020-04-05    5      4  2020     35       1             Afghanistan   
4    2020-04-04    4      4  2020      0       0             Afghanistan   
...         ...  ...    ...   ...    ...     ...                     ...   
9508 2020-03-25   25      3  2020      0       0                Zimbabwe   
9509 2020-03-24   24      3  2020      0       1                Zimbabwe   
9510 2020-03-23   23      3  2020      0       0                Zimbabwe   
9511 2020-03-22   22      3  2020      1       0                Zimbabwe   
9512 2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018 month_name   day_name  \
0       AF                  AFG   37172386.0      April  Wednesday   
1       AF                  AFG   37172386.0      April    Tuesday   
2       AF                  AFG   37172386.0      April     Monday   
3       AF                  AFG   37172386.0      April     Sunday   
4       AF                  AFG   37172386.0      April   Saturday   
...    ...                  ...          ...        ...        ...   
9508    ZW                  ZWE   14439018.0      March  Wednesday   
9509    ZW                  ZWE   14439018.0      March    Tuesday   
9510    ZW                  ZWE   14439018.0      March     Monday   
9511    ZW                  ZWE   14439018.0      March     Sunday   
9512    ZW                  ZWE   14439018.0      March   Saturday   

         pop_sum  death_5  
0     37172391.0        9  
1     37172391.0        5  
2     37172391.0        7  
3     37172391.0        6  
4     37172391.0        5  
...          ...      ...  
9508  14439023.0        5  
9509  14439023.0        6  
9510  14439023.0        5  
9511  14439023.0        5  
9512  14439023.0        5  

[9513 rows x 14 columns]

## idxmax()

### Q)  identify the row with the maximum number of deaths in your dataset

In [86]:
df.loc[df['deaths'].idxmax()]

dateRep                    2020-04-04 00:00:00
day                                          4
month                                        4
year                                      2020
cases                                     5233
deaths                                    2004
countriesAndTerritories                 France
geoId                                       FR
countryterritoryCode                       FRA
popData2018                         66987244.0
month_name                               April
day_name                              Saturday
pop_sum                             66987249.0
death_5                                   2009
Name: 3123, dtype: object

In [87]:
df['deaths'].max()

2004

In [89]:
df.groupby(['countriesAndTerritories'])['deaths'].max().idxmax()

'France'

In [91]:
df.groupby(['countriesAndTerritories'])['deaths'].max().sort_values(ascending=False)

countriesAndTerritories
France                      2004
United_States_of_America    1906
Italy                        971
Spain                        950
United_Kingdom               786
                            ... 
Papua_New_Guinea               0
South_Sudan                    0
Burundi                        0
Central_African_Republic       0
French_Polynesia               0
Name: deaths, Length: 204, dtype: int64

In [92]:
df.groupby(['countriesAndTerritories'])['deaths'].max().sort_values(ascending=False).idxmax()

'France'

### Q) Which country or territory has the lowest minimum death count in the dataset?

In [96]:
df.groupby(['countriesAndTerritories'])['deaths'].min().idxmin()

'Afghanistan'

### Q) Which row in the dataset corresponds to the lowest number of deaths, and what are the details of that row?

In [100]:
df.loc[df['deaths'].idxmin()]

dateRep                    2020-04-07 00:00:00
day                                          7
month                                        4
year                                      2020
cases                                       38
deaths                                       0
countriesAndTerritories            Afghanistan
geoId                                       AF
countryterritoryCode                       AFG
popData2018                         37172386.0
month_name                               April
day_name                               Tuesday
pop_sum                             37172391.0
death_5                                      5
Name: 1, dtype: object

### Q) Which month has the highest maximum number of cases recorded across all months in the dataset?

In [103]:
df.groupby(['month_name'])['cases'].max().idxmax()

'April'

## cut()

### Q) categorize the number of deaths in your dataset into different levels such as 'Low,' 'Medium,' 'High,' and 'Very High'

In [104]:
df['deaths'].max()

2004

In [105]:
df['deaths'].min()

0

In [107]:
df['death_category']=pd.cut(df['deaths'],bins=[-1,50,100,500,df['deaths'].max()],labels=['Low','Medium','High','Very High'])
# other methods
# 1) df['death_category']=pd.cut(df['deaths'],bins=[df['deaths'].min(),50,100,500,df['deaths'].max()],labels=['Low','Medium','High','Very High'])
# 2) df['death_category']=pd.cut(df['deaths'],bins=[-1,50,100,500,2004],labels=['Low','Medium','High','Very High'])

In [108]:
df

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0    2020-04-08    8      4  2020     30       4             Afghanistan   
1    2020-04-07    7      4  2020     38       0             Afghanistan   
2    2020-04-06    6      4  2020     29       2             Afghanistan   
3    2020-04-05    5      4  2020     35       1             Afghanistan   
4    2020-04-04    4      4  2020      0       0             Afghanistan   
...         ...  ...    ...   ...    ...     ...                     ...   
9508 2020-03-25   25      3  2020      0       0                Zimbabwe   
9509 2020-03-24   24      3  2020      0       1                Zimbabwe   
9510 2020-03-23   23      3  2020      0       0                Zimbabwe   
9511 2020-03-22   22      3  2020      1       0                Zimbabwe   
9512 2020-03-21   21      3  2020      1       0                Zimbabwe   

     geoId countryterritoryCode  popData2018 month_name   day_name  \
0       AF                  AFG   37172386.0      April  Wednesday   
1       AF                  AFG   37172386.0      April    Tuesday   
2       AF                  AFG   37172386.0      April     Monday   
3       AF                  AFG   37172386.0      April     Sunday   
4       AF                  AFG   37172386.0      April   Saturday   
...    ...                  ...          ...        ...        ...   
9508    ZW                  ZWE   14439018.0      March  Wednesday   
9509    ZW                  ZWE   14439018.0      March    Tuesday   
9510    ZW                  ZWE   14439018.0      March     Monday   
9511    ZW                  ZWE   14439018.0      March     Sunday   
9512    ZW                  ZWE   14439018.0      March   Saturday   

         pop_sum  death_5 death_category  
0     37172391.0        9            Low  
1     37172391.0        5            Low  
2     37172391.0        7            Low  
3     37172391.0        6            Low  
4     37172391.0        5            Low  
...          ...      ...            ...  
9508  14439023.0        5            Low  
9509  14439023.0        6            Low  
9510  14439023.0        5            Low  
9511  14439023.0        5            Low  
9512  14439023.0        5            Low  

[9513 rows x 15 columns]

In [109]:
df['death_category'].value_counts()

death_category
Low          9287
High          110
Medium         62
Very High      54
Name: count, dtype: int64